linkerd/identity/src/imp/openssl.rs

```diff
@@ -0,0 +1,345 @@
+                let client_config =
+                    ClientConfig::new(vec![], self.0.clone(), Some(crt.clone()), Some(key.clone()));
+
+                match verify {
+                    Ok(_) => Ok(CrtKey {
+                        id: crt.id.clone(),
+                        expiry: crt.expiry.clone(),
+                        client_config: Arc::new(client_config),
+                        server_config: Arc::new(server_config),
+                    }),
+                    Err(err) => Err(err),
+                }
+            }
+            Err(err) => Err(err.into()),
+        }
+    }
+
+    pub fn client_config(&self) -> Arc<ClientConfig> {
+        println!("Returning empty client config");
+        Arc::new(ClientConfig::new(vec![], self.0.clone(), None, None))
+    }
+}
+
+impl fmt::Debug for TrustAnchors {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "OpenSSL trust anchors")
+    }
+}
+
+#[derive(Clone, Debug)]
+pub enum InvalidCrt {
+    SubjectAltName(String),
+    Verify(X509VerifyResult),
+    General(Error),
+}
+
+impl From<Error> for InvalidCrt {
+    fn from(err: Error) -> Self {
+        InvalidCrt::General(err)
+    }
+}
+
+impl From<ErrorStack> for InvalidCrt {
+    fn from(err: ErrorStack) -> Self {
+        InvalidCrt::General(err.into())
+    }
+}
+
+impl fmt::Display for InvalidCrt {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            InvalidCrt::SubjectAltName(name) => write!(f, "Subject alt name incorrect {}", name),
+            InvalidCrt::Verify(err) => err.fmt(f),
+            InvalidCrt::General(err) => err.fmt(f),
+        }
+    }
+}
+
+impl error::Error for InvalidCrt {
+    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
+        match self {
+            InvalidCrt::Verify(err) => err.source(),
+            InvalidCrt::General(err) => err.source(),
+            _ => None,
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct CrtKey {
+    id: LocalId,
+    expiry: SystemTime,
+    client_config: Arc<ClientConfig>,
+    server_config: Arc<ServerConfig>,
+}
+
+// === CrtKey ===
+impl CrtKey {
+    pub fn name(&self) -> &Name {
+        self.id.as_ref()
+    }
+
+    pub fn expiry(&self) -> SystemTime {
+        self.expiry
+    }
+
+    pub fn id(&self) -> &LocalId {
+        &self.id
+    }
+
+    pub fn client_config(&self) -> Arc<ClientConfig> {
+        self.client_config.clone()
+    }
+
+    pub fn server_config(&self) -> Arc<ServerConfig> {
+        self.server_config.clone()
+    }
+}
+
+impl fmt::Debug for CrtKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> Result<(), fmt::Error> {
+        f.debug_struct("CrtKey")
+            .field("id", &self.id)
+            .field("expiry", &self.expiry)
+            .finish()
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct Crt {
+    pub(crate) id: LocalId,
+    expiry: SystemTime,
+    pub cert: X509,
+    pub chain: Vec<X509>,
```

Comment:
again, if the intention is to abstract over all the openssl-specific implementation details, we should ideally find a way to make these not be `pub`...

Reply:
Yeah I'm going to circle back to this.
The main thing here is the coupling between the Identity and the Tls modules respectively.
Ideally these should be completely hidden from the consumer. I'll figure something out in this regard :D